src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -353,35 +340,64 @@ public override void Flush()
             Interop.Termios.TermiosDiscard(_handle, Interop.Termios.Queue.AllQueues);
         }
 
-        public override long Seek(long offset, SeekOrigin origin)
+        internal int ReadByte(int timeout)
         {
-            throw new NotSupportedException(SR.NotSupported_UnseekableStream);
+            Read(_tempBuf, 0, 1, timeout);
+            return _tempBuf[0];
         }
 
-        public override void SetLength(long value)
+        public override int Read(byte[] array, int offset, int count)
         {
-            throw new NotSupportedException(SR.NotSupported_UnseekableStream);
+            return Read(array, offset, count, ReadTimeout);
         }
 
-        public override int ReadByte()
+        internal unsafe int Read(byte[] array, int offset, int count, int timeout)
         {
-            return ReadByte(ReadTimeout);
+            Task<int> t = ReadAsync(array, offset, count, GetCancellationTokenFromTimeout(timeout));
+
+            try
+            {
+                return t.Result;
```

Comment:
This should use `.GetAwaiter().GetResult()`... then you can just catch `OperationCanceledException rather than catching AggregateException and looking at its inner exception.

Separately, what's the default timeout?  Is it infinite?  It'd be really good if synchronous reads could actually be synchronous, rather than doing sync-over-async.

Reply:
> It'd be really good if synchronous reads could actually be synchronous, rather than doing sync-over-async.

Ah, the file handles are non-blocking.